Use TensorMethods.cpp

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* **#37639 Use TensorMethods.cpp**

Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

Commit:
Use TensorMethods.cpp

Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

dzhulgakov commented:
Hm, any concerns with overhead due to moving away from the inline methods?

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -339,16 +340,16 @@ class CAFFE2_API Tensor {
 
   template <typename T>
-  T * data_ptr() const;
+  TORCH_API T * data_ptr() const;
```
Hmm, that's weird. The `API` annotation on the class itself should have been sufficient. What's going on here?

ilia-cher replied:
err, will update; I was debugging a windows build error

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -661,6 +662,24 @@ class CAFFE2_API Tensor {
   c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_;
 };
 
+TORCH_API int64_t get_device(Tensor self);
```
Pretty sure this should be `CAFFE2_API`

ilia-cher replied:
as I understand `TORCH_API` is defined as `CAFFE2_API` in WindowsTorchApiMacro.h

ezyang reviewed (approved):
I'm a little skeptical about the `_API` on the methods, but the rest looks ok

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

ilia-cher commented:
> Hm, any concerns with overhead due to moving away from the inline methods?

will run Sebastian's model benchmark and post here

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

ezyang commented:
@dzhulgakov: @malfet was mentioning to me about how these methods were so large already, they probably weren't getting inlined anyway XD

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

ilia-cher commented:
fixing ::data_ptr and ::item visibility

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

Commit:
Update on "Use TensorMethods.cpp"


Summary:
Changing TensorMethods.h to .cpp
Necessary to avoid incomplete types in dispatcher

Test Plan:
CI

Differential Revision: [D21374247](https://our.internmc.facebook.com/intern/diff/D21374247)

[ghstack-poisoned]

ezyang commented:
This broke mobile code analysis. It will not be too easy to fix, we should talk about this with @ljk53 in composability.

ilia-cher commented:
> This broke mobile code analysis. It will not be too easy to fix, we should talk about this with @ljk53 in composability.

@ezyang @ljk53  why was all of CI green? how can I repro?

ljk53 commented:
> mobile

It didn't break because the CI only runs on landed PR. I made it so because there is another e2e CI also covering most part of mobile code analysis. There is only a small unit test that e2e CI doesn't run, which caused the failure.

This is the error message on failed CI job:
```
Jun 01 15:31:43 In file included from /var/lib/jenkins/workspace/test/mobile/op_deps/simple_ops.cpp:1:
Jun 01 15:31:43 /var/lib/jenkins/workspace/test/mobile/op_deps/simple_ops.h:22:52: error: no member named 'singleton' in 'c10::DispatchKey'
Jun 01 15:31:43   static c10::OperatorHandle op = c10::Dispatcher::singleton()
Jun 01 15:31:43                                   ~~~~~~~~~~~~~~~~~^
Jun 01 15:31:43 /var/lib/jenkins/workspace/test/mobile/op_deps/simple_ops.h:24:15: error: no member named 'Dispatcher' in namespace 'c10'; did you mean 'DispatchKey'?
Jun 01 15:31:43   return c10::Dispatcher::singleton().call<Tensor, const Tensor&>(
Jun 01 15:31:43          ~~~~~^~~~~~~~~~
Jun 01 15:31:43               DispatchKey
```

I think we need change it accordingly: https://github.com/pytorch/pytorch/blob/master/test/mobile/op_deps/simple_ops.h

You can repeat the test locally using:
```
https://github.com/pytorch/pytorch/blob/master/tools/code_analyzer/build.sh#L12
```

ezyang commented:
You can force all of master ci to run by pushing to ci-all/blah branch on pytorch/pytorch
